### PR TITLE
Mots 264: Added pagination and search of locations

### PR DESCRIPTION
**Pull request by pstudzinski**
Added pagination and search of locations.
Also added constans of parameters.
https://applab.atlassian.net/browse/MOTS-264

### CODE REVIEW

File: src/main/java/org/motechproject/mots/web/LocationController.java [hunk 1]
```diff
@@ -90,4 +100,85 @@
...
+
+  /**
+   * Finds chiefdoms matching all of the provided parameters.
+   * If there are no parameters, return all chiefdoms.
+   */
+  @RequestMapping(value = "chiefdom/locations/search", method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<LocationPreviewDto> searchChiefdoms(
+      @RequestParam(value = NAME_PARAM, required = false) String chiefdomName,
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/LocationController.java]: Should be "name"

File: src/main/java/org/motechproject/mots/web/LocationController.java [hunk 2]
```diff
@@ -90,4 +100,85 @@
...
+
+  /**
+   * Finds districts matching all of the provided parameters.
+   * If there are no parameters, return all districts.
+   */
+  @RequestMapping(value = "district/locations/search", method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<LocationPreviewDto> searchDistricts(
+      @RequestParam(value = NAME_PARAM, required = false) String districtName,
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/LocationController.java]: Should be "name"

File: src/main/java/org/motechproject/mots/web/LocationController.java [hunk 3]
```diff
@@ -90,4 +100,85 @@
...
+
+  /**
+   * Finds communities matching all of the provided parameters.
+   * If there are no parameters, return all communities.
+   */
+  @RequestMapping(value = "community/locations/search", method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<LocationPreviewDto> searchCommunities(
+      @RequestParam(value = NAME_PARAM, required = false) String communityName,
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/LocationController.java]: Should be "name"

File: src/main/java/org/motechproject/mots/web/LocationController.java [hunk 4]
```diff
@@ -90,4 +100,85 @@
...
+  /**
+   * Finds facilities matching all of the provided parameters.
+   * If there are no parameters, return all facilities.
+   */
+  @RequestMapping(value = "facility/locations/search", method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<LocationPreviewDto> searchFacilities(
+      @RequestParam(value = FACILITY_ID_PARAM, required = false) String facilityId,
+      @RequestParam(value = NAME_PARAM, required = false) String facilityName,
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/LocationController.java]: Should be "name"

File: src/main/java/org/motechproject/mots/web/LocationController.java [hunk 5]
```diff
@@ -90,4 +100,85 @@
...
+
+  /**
+   * Finds facilities matching all of the provided parameters.
+   * If there are no parameters, return all facilities.
+   */
+  @RequestMapping(value = "facility/locations/search", method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<LocationPreviewDto> searchFacilities(
+      @RequestParam(value = FACILITY_ID_PARAM, required = false) String facilityId,
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/LocationController.java]: Should be "id"

File: src/main/java/org/motechproject/mots/web/LocationController.java [hunk 6]
```diff
@@ -90,4 +100,85 @@
...
+      @RequestParam(value = NAME_PARAM, required = false) String facilityName,
+      @RequestParam(value = FACILITY_TYPE_PARAM, required = false) String facilityType,
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/LocationController.java]: Should be "type"

File: src/main/java/org/motechproject/mots/web/LocationController.java (hunk 5 above)
Comment (kkaczmarczyk) [reply src/main/java/org/motechproject/mots/web/LocationController.java]: it's ok actually

File: src/main/java/org/motechproject/mots/web/LocationController.java (hunk 6 above)
Comment (kkaczmarczyk) [reply src/main/java/org/motechproject/mots/web/LocationController.java]: it's ok actually

File: src/main/java/org/motechproject/mots/repository/custom/impl/BaseRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  protected static final String DISTRICT = "district";
+  protected static final String COMMUNITY = "community";
+  protected static final String INCHARGE = "incharge";
+  protected static final String EMAIL = "email";
+
+  @PersistenceContext
+  protected EntityManager entityManager;
+
+  protected static Pair<Integer, Integer> getMaxAndFirstResult(Pageable pageable) {
+    int pageSize = null != pageable ? pageable.getPageSize() : 0;
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/repository/custom/impl/BaseRepositoryImpl.java]: let's have separate getPageSize and getFirstResult methods

File: src/main/java/org/motechproject/mots/repository/custom/impl/CommunityRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+    List<Order> orders = new ArrayList<>();
+    Iterator<Sort.Order> iterator = pageable.getSort().iterator();
+
+    Sort.Order order;
+    Path<T> path;
+    while (iterator.hasNext()) {
+      order = iterator.next();
+      if (order.getProperty().equals(LocationController.PARENT_PARAM)) {
+        path = root.get(FACILITY).get(NAME);
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/repository/custom/impl/CommunityRepositoryImpl.java]: maybe it's worth creating getPath() method and using addSortProperties() of base class

Comment (pstudzinski) [reply src/main/java/org/motechproject/mots/repository/custom/impl/CommunityRepositoryImpl.java]: I changed.

**Pull request merged**